src/adm/filter.rs

```diff
@@ -112,41 +112,41 @@ pub fn spawn_updater(
             return Ok(());
         }
     }
-    let mfilter = filter.clone();
+    let mfilter = Arc::clone(filter);
     rt::spawn(async move {
-        let mut tags = crate::tags::Tags::default();
         loop {
-            {
-                // Do the check inside of a scope so that the read lock can be released right away.
-                let should_update = { mfilter.read().await.requires_update(&storage_client).await };
-                match should_update {
-                    Ok(true) => {
-                        let mut filter = mfilter.write().await;
-                        match filter.update(&storage_client).await {
-                            Ok(_) => {
-                                metrics.incr("filter.adm.update.ok").ok();
-                            }
-                            Err(e) => {
-                                filter.report(&e, &mut tags);
-                                metrics.incr("filter.adm.update.error").ok();
-                            }
-                        }
-                    }
-                    Ok(false) => {
-                        metrics.incr("filter.adm.update.check.skip").ok();
-                    }
-                    Err(e) => {
-                        mfilter.read().await.report(&e, &mut tags);
-                        metrics.incr("filter.adm.update.check.error").ok();
-                    }
-                }
-            }
+            updater(&mfilter, &storage_client, &metrics).await;
             rt::time::sleep(refresh_rate).await;
         }
     });
     Ok(())
 }
 
+/// Update `AdmFilter` from the Cloud Storage settings if they've been updated
+async fn updater(
+    filter: &Arc<RwLock<AdmFilter>>,
+    storage_client: &cloud_storage::Client,
+    metrics: &Arc<StatsdClient>,
+) {
+    // Do the check before matching so that the read lock can be released right away.
+    let result = filter.read().await.fetch_new_settings(storage_client).await;
```

Comment:
Ah, I didn't know that calling `fetch_new_settings(&self)` would also drop the lock guard from the current scope. TIL!